packages/xml-views-tooltip/src/tooltip.ts

```diff
@@ -0,0 +1,140 @@
+import { XMLAttribute, XMLElement, DEFAULT_NS } from "@xml-tools/ast";
+import { isXMLNamespaceKey } from "@xml-tools/common";
+import {
+  XMLElementOpenName,
+  XMLElementCloseName,
+  XMLAttributeKey,
+  XMLAttributeValue,
+} from "@xml-tools/ast-position";
+import { find } from "lodash";
+import {
+  xmlToFQN,
+  ui5NodeToFQN,
+  flattenAggregations,
+  getUI5ClassByXMLElement,
+  getUI5PropertyByXMLAttributeKey,
+} from "@ui5-language-assistant/logic-utils";
+import {
+  UI5Class,
+  UI5Aggregation,
+  UI5SemanticModel,
+  UI5EnumValue,
+  UI5Enum,
+  BaseUI5Node,
+} from "@ui5-language-assistant/semantic-model-types";
+import { findSymbol } from "@ui5-language-assistant/semantic-model";
+
+export function findUI5HoverNodeAtOffset(
+  visitor:
+    | XMLElementOpenName
+    | XMLElementCloseName
+    | XMLAttributeKey
+    | XMLAttributeValue
+    | undefined,
+  model: UI5SemanticModel
+): BaseUI5Node | undefined {
+  if (visitor === undefined) {
+    return undefined;
+  }
+  switch (visitor.kind) {
+    case "XMLElementOpenName":
+      return getUI5NodeByElement(visitor.astNode, model, visitor.kind);
+    case "XMLElementCloseName":
+      return getUI5NodeByElement(visitor.astNode, model, visitor.kind);
+    case "XMLAttributeKey":
+      return getUI5PropertyByXMLAttributeKey(visitor.astNode, model);
+    case "XMLAttributeValue":
+      return findUI5NodeByXMLAttributeValue(visitor.astNode, model);
```

Comment:
Add `assertNever` in the default case (so we don't forget to update this function if we support more types)